### PR TITLE
Migrate to Jest

### LINKED ISSUE #53: Migrate tests to Jest
Replace `tape` with `jest` per the MetaMask module template, and colocate tests with source files.

### PR DESCRIPTION
**Pull request by FrederikBolding**
Migrates to use Jest instead of Tape

Fixes https://github.com/MetaMask/eth-rpc-errors/issues/53

### CODE REVIEW

File: src/errors.test.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+const dummyData = { foo: 'bar' };
+
+const SERVER_ERROR_CODE = -32098;
+const CUSTOM_ERROR_CODE = 1001;
+const CUSTOM_ERROR_MESSAGE = 'foo';
+
+describe('ensure exported object accepts a single string argument where appropriate', () => {
+  let err = ethErrors.rpc.invalidInput(CUSTOM_ERROR_MESSAGE);
+  expect(err.code === errorCodes.rpc.invalidInput).toBeTruthy();
```
Comment (Mrtenz) [inline src/errors.test.ts]: `expect` should not be nested inside `describe` directly.

File: src/errors.test.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  expect(err.message === CUSTOM_ERROR_MESSAGE).toBeTruthy();
+
+  err = ethErrors.provider.unauthorized(CUSTOM_ERROR_MESSAGE);
+  expect(err.code === errorCodes.provider.unauthorized).toBeTruthy();
+  expect(err.message === CUSTOM_ERROR_MESSAGE).toBeTruthy();
```
Comment (Mrtenz) [inline src/errors.test.ts]: These tests can be changed to something like `expect(err.message).toBe(CUSTOM_ERROR_MESSAGE);`. That would be easier to debug on potential test failures too.

File: src/errors.test.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ethErrors.provider.custom('bar' as any);
+  }).toThrowError(/Ethereum Provider custom errors must/u);
+
+  expect(() => {
+    ethErrors.provider.custom({ code: 4009, message: 2 } as any);
+  }).toThrowError(/"message" must be/u);
+
+  expect(() => {
+    ethErrors.provider.custom({ code: 4009, message: '' });
+  }).toThrowError(/"message" must be/u);
```
Comment (Mrtenz) [inline src/errors.test.ts]: Can we test for full error messages, rather than a partial RegEx?

File: src/errors.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+  const err = ethErrors.provider.custom({ code: 4009, message: 'foo' });
+  expect(err instanceof EthereumProviderError).toBeTruthy();
+});
+
+describe('server rpc error options', () => {
+  expect(() => {
+    ethErrors.rpc.server('bar' as any);
```
Comment (Mrtenz) [inline src/errors.test.ts]: `@ts-expect-error`.

File: src/errors.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  const err = ethErrors.rpc.server({ code: -32006, message: 'foo' });
+  expect(err instanceof EthereumRpcError).toBeTruthy();
+});
+
+// we just iterate over all keys on the errors object and call and check
+// each error in turn
+describe('test exported object for correctness', () => {
+  it.each(Object.entries(ethErrors.rpc).filter(([key]) => key !== 'server'))(
+    'Ethereum RPC',
```
Comment (Mrtenz) [inline src/errors.test.ts]: This should describe what is being tested.

File: src/errors.test.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    (key, value) => {
+      const createError = value as any;
+      const error = createError({
+        message: null,
+        data: Object.assign({}, dummyData),
+      });
+      // @ts-expect-error Fix?
+      const rpcCode = rpcCodes[key];
```
Comment (Mrtenz) [inline src/errors.test.ts]: ?

File: src/errors.test.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      expect(
+        rpcCodeValues.includes(error.code) ||
+          (error.code <= -32000 && error.code >= -32099),
+      ).toBeTruthy();
+      expect(error.code).toBe(rpcCode);
+      expect(error.message).toBe(getMessageFromCode(rpcCode));
+    },
+  );
+
+  it('server', () => {
```
Comment (Mrtenz) [inline src/errors.test.ts]: This should describe what is being tested.

File: src/errors.test.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      message: undefined,
+      data: Object.assign({}, dummyData),
+    });
+    expect(error.code <= -32000 && error.code >= -32099).toBeTruthy();
+    expect(error.message).toBe(serverErrorMessage);
+  });
+
+  it.each(
+    Object.entries(ethErrors.provider).filter(([key]) => key !== 'custom'),
+  )('Ethereum Provider', (key, value) => {
```
Comment (Mrtenz) [inline src/errors.test.ts]: This should describe what is being tested.

File: src/errors.test.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      data: Object.assign({}, dummyData),
+    });
+    // @ts-expect-error Fix?
+    const providerCode = providerCodes[key];
+    expect(error.code >= 1000 && error.code < 5000).toBeTruthy();
+    expect(error.code).toBe(providerCode);
+    expect(error.message).toBe(getMessageFromCode(providerCode));
+  });
+
+  it('custom', () => {
```
Comment (Mrtenz) [inline src/errors.test.ts]: This should describe what is being tested.

File: src/serializeError.test.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+};
+const validError2 = ethErrors.rpc.parse();
+delete validError2.stack;
+const validError3 = ethErrors.rpc.parse(dummyMessage);
+delete validError3.stack;
+const validError4 = ethErrors.rpc.parse({
+  message: dummyMessage,
+  data: Object.assign({}, dummyData),
+});
+delete validError4.stack;
```
Comment (Mrtenz) [inline src/serializeError.test.ts]: Can we move this to a separate file?

File: src/errors.test.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+const providerCodes = errorCodes.provider;
+const serverErrorMessage = JSON_RPC_SERVER_ERROR_MESSAGE;
+
+const rpcCodeValues = Object.values(rpcCodes);
+
+const dummyData = { foo: 'bar' };
+
+const SERVER_ERROR_CODE = -32098;
+const CUSTOM_ERROR_CODE = 1001;
+const CUSTOM_ERROR_MESSAGE = 'foo';
```
Comment (Mrtenz) [inline src/errors.test.ts]: Can we move this to a separate file?

File: src/serializeError.test.ts [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+describe('serializeError', () => {
+  it('invalid error: non-object', () => {
```
Comment (Mrtenz) [inline src/serializeError.test.ts]: These tests can probably be combined with `it.each`.

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: src/errors.test.ts (hunk 6 above)
Comment (FrederikBolding) [reply src/errors.test.ts]: Explained now

Comment (FrederikBolding) [reply src/errors.test.ts]: Not sure we can fix

File: src/serializeError.test.ts (hunk 12 above)
Comment (FrederikBolding) [reply src/serializeError.test.ts]: Not sure it is worth it to spend more time on it

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**